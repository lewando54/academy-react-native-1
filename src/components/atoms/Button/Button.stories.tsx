/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { StoryObj, Meta } from '@storybook/react'

import Button from './Button'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
	component: Button,
	// More on argTypes: https://storybook.js.org/docs/react/api/argtypes
	argTypes: {
		color: { control: 'select', options: ['primary', 'secondary'] },
		onClick: { action: 'clicked' }
	}
} satisfies Meta<typeof Button>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {
// More on args: https://storybook.js.org/docs/react/writing-stories/args
	args: {
		color: 'primary',
		children: 'Button'
	}
}

export const Secondary: Story = {
	args: {
		color: 'secondary',
		children: 'Button'
	}
}