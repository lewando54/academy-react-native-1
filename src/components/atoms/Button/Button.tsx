import React, { useRef } from 'react'
import { Pressable, Text, GestureResponderEvent, Animated } from 'react-native'
import buttonStyle from './Button.style'

type colorProp = 'primary' | 'secondary' | string

interface ButtonProps {
  children: string
  color?: colorProp
  onClick?: ((event: GestureResponderEvent) => void) | null | undefined
  testId?: string
}

function Button({ color = 'primary', children, onClick, testId }: ButtonProps): React.ReactElement{

	const animatedButtonScale = useRef(new Animated.Value(1)).current

	const onPressIn = () => {
		Animated.spring(animatedButtonScale, {
			toValue: 1.25,
			speed: 200,
			useNativeDriver: true,
		}).start()
	}
  
	const onPressOut = () => {
		Animated.spring(animatedButtonScale, {
			toValue: 1,
			speed: 200,
			useNativeDriver: true,
		}).start()
	}

	let selectedColor = buttonStyle.primary
	if(color === 'secondary'){
		selectedColor = buttonStyle.secondary
	}
	else if (color !== 'primary'){
		selectedColor = {backgroundColor: color}
	}

	return (
		<Animated.View style={{transform: [{scale: animatedButtonScale}], width: '100%'}}>
			<Pressable style={[buttonStyle.base, selectedColor]} onPress={onClick} onPressIn={onPressIn} onPressOut={onPressOut} testID={testId}>
				<Text style={buttonStyle.text}>{children}</Text>
			</Pressable>
		</Animated.View>
	)
}

export default Button