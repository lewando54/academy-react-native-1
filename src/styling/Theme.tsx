export const PRIMARY_COLOR_BUTTON = 'rgb(236, 72, 153)'
export const SECONDARY_COLOR_BUTTON = 'rgb(229, 231, 235)'