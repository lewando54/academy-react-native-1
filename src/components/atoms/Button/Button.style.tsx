import { StyleSheet } from 'react-native'
import { PRIMARY_COLOR_BUTTON, SECONDARY_COLOR_BUTTON } from '../../../styling/Theme'

export default StyleSheet.create({
    base: {
        borderRadius: 6,
        width: '100%',
        paddingTop: 8,
        paddingBottom: 8,
        paddingHorizontal: 4,
        shadowOffset: {width: 0, height: 4},
        shadowBlur: 6,
        shadowColor: 'rgba(0,0,0,0.3)',
        border: 0,
    },
    text:{
        color: 'white',
        fontSize: 14,
        textTransform: 'uppercase',
        letterSpacing: 0.4,
        lineHeight: 20,
        textAlign: 'center'
    },
    primary: {
        backgroundColor: PRIMARY_COLOR_BUTTON,
    },
    secondary: {
        backgroundColor: SECONDARY_COLOR_BUTTON,
    }
})