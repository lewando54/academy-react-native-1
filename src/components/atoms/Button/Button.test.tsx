import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react-native'
import { PRIMARY_COLOR_BUTTON, SECONDARY_COLOR_BUTTON } from '../../../styling/Theme'
import Button from './Button'

describe('Button', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Button testId='testxd' color="primary">Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render with primary color by default', () => {
    const tree = renderer.create(<Button testId='testxd'>Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render innerHTML text', () => {
    const { getByText } = render(<Button testId='testxd' color="secondary">Click me</Button>)
    expect(getByText('Click me')).toBeTruthy()
  })

  it('should render with primary color', () => {
    const { getByTestId } = render(<Button testId='button' color={'primary'}>Click me</Button>)
    expect( getByTestId('button').props.style[1].backgroundColor).toEqual(PRIMARY_COLOR_BUTTON)
  })

  it('should render with secondary color', () => {
    render(<Button testId='button' color={'secondary'}>Click me</Button>)
    expect(screen.getByTestId('button').props.style[1].backgroundColor).toEqual(SECONDARY_COLOR_BUTTON)
  })

  it('should render with custom color', () => {
    render(<Button testId='button' color={'#fff'}>Click me</Button>)
    expect(screen.getByTestId('button').props.style[1].backgroundColor).toEqual('#fff')
  })

  it('should call onClick function when clicked', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(<Button testId='testxd' onClick={onClick}>Click me</Button>)
    fireEvent.press(getByTestId('testxd'))
    expect(onClick).toHaveBeenCalled()
  })

})
