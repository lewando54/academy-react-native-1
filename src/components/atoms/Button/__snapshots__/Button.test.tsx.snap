// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "transform": [
        {
          "scale": 1,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "border": 0,
          "borderRadius": 6,
          "paddingBottom": 8,
          "paddingHorizontal": 4,
          "paddingTop": 8,
          "shadowBlur": 6,
          "shadowColor": "rgba(0,0,0,0.3)",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "width": "100%",
        },
        {
          "backgroundColor": "rgb(236, 72, 153)",
        },
      ]
    }
    testID="testxd"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 14,
          "letterSpacing": 0.4,
          "lineHeight": 20,
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    >
      Click me
    </Text>
  </View>
</View>
`;

exports[`Button should render with primary color by default 1`] = `
<View
  collapsable={false}
  style={
    {
      "transform": [
        {
          "scale": 1,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "border": 0,
          "borderRadius": 6,
          "paddingBottom": 8,
          "paddingHorizontal": 4,
          "paddingTop": 8,
          "shadowBlur": 6,
          "shadowColor": "rgba(0,0,0,0.3)",
          "shadowOffset": {
            "height": 4,
            "width": 0,
          },
          "width": "100%",
        },
        {
          "backgroundColor": "rgb(236, 72, 153)",
        },
      ]
    }
    testID="testxd"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 14,
          "letterSpacing": 0.4,
          "lineHeight": 20,
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    >
      Click me
    </Text>
  </View>
</View>
`;
